tests/consensus_tests/test_shard_transfer_rate_limiting.py

```diff
@@ -0,0 +1,245 @@
+                 "payload": {"count": [0]}},
+                {"id": 6, "vector": [0.35, 0.08, 0.11, 0.44]}
+            ]
+        })
+    assert_http_ok(r)
+
+    # Check that 'search' returns the same results on all peers
+    for uri in peer_api_uris:
+        r = requests.post(
+            f"{uri}/collections/test_collection/points/search", json={
+                "vector": [0.2, 0.1, 0.9, 0.7],
+                "top": 3,
+            }
+        )
+        assert_http_ok(r)
+        assert r.json()["result"][0]["id"] == 4
+        assert r.json()["result"][1]["id"] == 1
+        assert r.json()["result"][2]["id"] == 3
+
+    # Extract current collection cluster info
+    collection_cluster_info = get_collection_cluster_info(peer_api_uris[0], "test_collection")
+    target_peer_id = collection_cluster_info["remote_shards"][0]["peer_id"]
+    source_uri = peer_api_uris[0]
+    target_uri = peer_api_uris[1]
+
+    target_collection_cluster_info = get_collection_cluster_info(target_uri, "test_collection")
+    target_before_local_shard_count = len(target_collection_cluster_info["local_shards"])
+
+    before_local_shard_count = len(collection_cluster_info["local_shards"])
+    shard_id = collection_cluster_info["local_shards"][0]["shard_id"]
+    source_peer_id = collection_cluster_info["peer_id"]
+
+    # set rate limiting
+    r = requests.patch(
+        f"{source_uri}/collections/test_collection", json={
+            "strict_mode_config": {
+                "enabled": True,
+                "write_rate_limit": 1,
+                # TODO(ratelimit) validate read limits as well "read_rate_limit": 1
+            }
+        })
+    assert_http_ok(r)
+
+    # Exhaust write rate limit on target shard
+    rate_limited = False
+    for _ in range(100):
+        r = requests.put(
+            f"{target_uri}/collections/test_collection/points?wait=true", json={
+                "points": [
+                    {
+                        "id": 1,
+                        "vector": [0.05, 0.61, 0.76, 0.74],
+                        "payload": {
+                            "city": "Berlin",
+                            "country": "Germany",
+                            "count": 1000000,
+                            "square": 12.5,
+                            "coords": {"lat": 1.0, "lon": 2.0}
+                        }
+                    },
+                ]
+            })
+        if not r.ok:
+            rate_limited = True
+            assert r.status_code == 429, f"Failure body {r.json()}"
+            assert "Rate limiting exceeded: Write rate limit exceeded, retry later" in r.json()['status']['error']
+            break
+
+    assert rate_limited, "Expected write rate limit to be reached"
+
+    # Move shard `shard_id` to peer `target_peer_id` without write budget
+    r = requests.post(
+        f"{source_uri}/collections/test_collection/cluster", json={
+            "move_shard": {
```

Comment:
Just to be clear, move shard does not recreate the entire replica on the target peer and keeps the existing rate limiter in place?

I'm asking because some of our shard operations may recreate the replica, in which case it may also recreate and replenish the rate limiter.

Reply:
Move sets up the the local shard in the existing target peer's replica set.

But we track only operations done to the local shard anyway, initializing a new local shard will always result in a new full rate limiter.